src/iface/ethernet.rs

```diff
@@ -990,4 +1011,56 @@ mod test {
             &IpAddress::Ipv4(remote_ip_addr)),
             Ok((remote_hw_addr, MockTxToken)));
     }
+
+    #[test]
+    #[cfg(feature = "proto-icmpv4")]
+    fn test_icmpv4_socket() {
+        use socket::{Icmpv4PacketBuffer, Icmpv4Socket, Icmpv4SocketBuffer};
+        use wire::{Icmpv4Repr, Icmpv4Packet};
+
+        let (iface, mut socket_set) = create_loopback();
+
+        let rx_buffer = Icmpv4SocketBuffer::new(vec![Icmpv4PacketBuffer::new(vec![0; 20])]);
+        let tx_buffer = Icmpv4SocketBuffer::new(vec![Icmpv4PacketBuffer::new(vec![0; 20])]);
+
+        let mut icmp_bytes = vec![0u8; 20];
+        let mut icmp_packet = Icmpv4Packet::new(&mut icmp_bytes);
+
+        let icmpv4_socket = Icmpv4Socket::new(rx_buffer, tx_buffer);
+
+        let socket_handle = socket_set.add(icmpv4_socket);
+
+        {
+            let mut icmp_socket = socket_set.get::<Icmpv4Socket>(socket_handle);
+
+            assert_eq!(icmp_socket.recv(Ipv4Address([0x7f, 0x00, 0x00, 0x02])),
+                       Err(Error::Exhausted));
```

Comment:
This is what I really don't like about this implementation. I originally wanted to stay away from using a `bind` call, but now as a result if you need to use a `recv` as the `bind` in a way.

Reply:
I think this will need to be reworked to have a proper `bind` call, at least if we want to support downstreams like redox which present a POSIX-like interface. Take a look at this: https://www.planet-lab.org/raw_sockets/api_icmp.html.

Reply:
That is a super helpful doc